### PR TITLE
Migrate code into Typescript

### PR DESCRIPTION
**Pull request by Jianrong-Yu**
Similar to #169, I tried to migrate everything to TypeScript in this PR.

Some changes:
* `Vite` is replaced to `tsc` to generate both ESM and CommonJS package

### CODE REVIEW

**Comment by crecotun** [comment]
Hi @Jianrong-Yu 
Thank you for doing this.
Would you mind making sure this is going to be marked as Breaking Change and result in a major version update?

**Comment by tzimmermann** [comment]
Just published this as `5.0.0-rc1` under `next` tag on NPM: https://www.npmjs.com/package/react-stick/v/5.0.0-rc1
To test in your app, use `npm install --save react-stick@next`

**Comment by tzimmermann** [comment]
I've checked so far

- Prettier & Eslint commit hooks are working as expected.
- Usage of 5.0.0-rc1 in our app (pex-client) and in glucose works as expected.
- Compared the sample HTML pages and could not spot any differences to master.

What I haven't checked:
- All the JS->TS transformations (I am no TS expert). Does it even make sense to review them, or did you generate them with flowts @Jianrong-Yu ?

**Comment by crecotun** [comment]
I am not sure about replacing Vite with `tsc`.
My idea was to move our demos from webpack to Vite.
Vite does support TypeScript out of the box, so replacing the existing functionality simply doesn't bring any value.
I'd consider reverting this change.
What do you think @tzimmermann @Jianrong-Yu ?

**Review by crecotun — CHANGES_REQUESTED** [review]

**Comment by Jianrong-Yu** [comment]
> What I haven't checked:
> 
> * All the JS->TS transformations (I am no TS expert). Does it even make sense to review them, or did you generate them with flowts @Jianrong-Yu ?

Hi @tzimmermann Yes the type is guaranteed by `flowts`, the only change on type is the `StickPropsT` now have also all the attributes of an HTMLElement (like `onMouseEnter` and `onMouseLeave` used in `demo/src/regressions/StickOnHover.tsx`). It's because I found the `...rest` always passed to the container:
```
    <Component
      {...rest}
      {...styles}
```

> Hi @Jianrong-Yu Thank you for doing this. Would you mind making sure this is going to be marked as Breaking Change and result in a major version update?

Hi @crecotun The API didn't change, so from the aspect of users there is no breaking change for `react-stick`.


> I am not sure about replacing Vite with `tsc`. My idea was to move our demos from webpack to Vite. Vite does support TypeScript out of the box, so replacing the existing functionality simply doesn't bring any value. I'd consider reverting this change. What do you think @tzimmermann @Jianrong-Yu ?

* Bundling: I tried Typescript in `Vite` before I replaced it with `tsc`. Vite itself will not do the bundling for package; the bundling is done by `rollup`. To make the `rollup` working for typescript, plugin `@rollup/plugin-typescript` must be installed and put in the Vite configuration. Since it can be done just by using `tsc`, why choose another tool wrapped **twice** to do almose the same thing?
* Testing: Testing must be done by `Karma` since it's the best choice for the unit-tests in a real browser. `Jest` is not working well in this case. And `webpack` with `babel` is the only combination I tried successfully on TypeScript/Coverage  support in `Karma`, so `webpack` and `babel` (used in `tests/karma.conf.js`) are also required.
* Demo: Since the `webpack` is required by `testing`, I don't think it's a good idea to migrate Demo into Vite. We always need webpack so why not reuse it in Demo?

File: .vscode/extensions.json
Comment (Jianrong-Yu) [reply .vscode/extensions.json:1]: Why? It makes onboarding the new developer easier.
VSCode will recommend the developer to install the extensions which make the development easier. If the extensions are already installed, it will not bother the user.

File: .vscode/settings.json
Comment (Jianrong-Yu) [reply .vscode/settings.json:11]: You mean `editor.formatOnSave` one? Or `prettier` config moved to `eslint`? 
We can remove `formatOnSave` if you don't like it. The format will be fixed by `lint-staged` when the code is committing anyway.

Comment (crecotun) [reply .vscode/settings.json:11]: Sorry, I should provide more context here.

All the editor settings that you're adding will change the developer's IDE normal behavior and I usually tend to avoid doing this as everyone has their preferences.

What we can do and we already are doing — add some pre-commit hooks to ensure code passes eslint and prettier.

And it's not fair to add settings for vscode, and completely ignore WebStorm, for example. Having settings for all IDE will just pollute the project.

What do you think?

**Comment by crecotun** [comment]
@Jianrong-Yu I wouldn't agree that for users there was no change, you replaced Flow with TypeScript, so end-user is affected and this is a major change. I might be wrong here.

Regarding the webpack.
Yes, it does make sense. I wasn't aware that for making it work with rollup we have to install an additional plugin.
So I'd say yes, we can live with webpack and `tsc` compiler as long as they do their job.

Testing.
One thing to mention here, I'd consider moving from Karma to Cypress + Jest if that makes sense. Cypress allows testing components in isolation.

**Comment by Jianrong-Yu** [comment]
> Sorry, I should provide more context here.
> 
> All the editor settings that you're adding will change the developer's IDE normal behavior and I usually tend to avoid doing this as everyone has their preferences.
> 
> What we can do and we already are doing — add some pre-commit hooks to ensure code passes eslint and prettier.
> 
> And it's not fair to add settings for vscode, and completely ignore WebStorm, for example. Having settings for all IDE will just pollute the project.
> 
> What do you think?

@crecotun I agree it's not fair to have settings for vscode only, but since vscode is almost the de facto editor in front-end community, I think it's good to have the vscode only settings. The Typescript is also opinionated, but it's also the de facto javascript static typing solution now: the developers are still able to use `react-stick` in Javascript, but they will get better support if they are using Typescript; just like users of VSCode will get more support, but the users of other editor will still get the formatting by Husky and Lint-staged in committing. 
BTW, the vscode `settings.json` is already there before this PR, I just added a few new items in it.

**Review by crecotun — DISMISSED** [review]

**Comment by Jianrong-Yu** [comment]
@crecotun @tzimmermann `commitlint` is added [in these new commits](https://github.com/signavio/react-stick/pull/411/files/3695495cfdc4f7af9d268d08e2cfaae6df1b518f..8c6533d78487621aff2f7c4559664c5f739fe513), can you do a review on them? Thanks.

BTW should I merge this PR once it's approved? The version number will be jump to 5.0 since it's a breaking change on typing.

**Comment by crecotun** [comment]
@Jianrong-Yu Could you explain why you added commitlint to the circleci?

**Comment by Jianrong-Yu** [comment]
> @Jianrong-Yu Could you explain why you added commitlint to the circleci?

@crecotun I found all my old commits did't meet the requirements of `semantic-release` but it passed the circleCI check, I think it's better to have this check in CI so the developer will get notification for the comment if it's not met the requirement. The `husky` with `lint-staged` is useful for developers to push changes on local env but it cannot notify the user for any change done on Github web site directly.
If you think it's a bit overkill, I'll revert the change on circleCI.

**Comment by crecotun** [comment]
@Jianrong-Yu I'd rather keep this pr about TS only and if you don't mind, create another PR with commit lint improvements.
Would that work for you?

**Pull request merged**
